ably/realtime_conn.go

```diff
@@ -958,8 +958,10 @@ func (c *Connection) lockSetState(state ConnectionState, err error, retryIn time
 func (c *Connection) ctxCancelOnStateTransition() (context.Context, context.CancelFunc) {
 	ctx, cancel := context.WithCancel(context.Background())
 
-	off := c.internalEmitter.OnceAll(func(ConnectionStateChange) {
-		cancel()
+	off := c.internalEmitter.OnceAll(func(change ConnectionStateChange) {
+		if change.Current == ConnectionStateConnected || change.Current == ConnectionStateClosed {
```

Comment:
The purpose of this function is to stop the timer that moves the connection state from DISCONNECTED or SUSPENDED to CONNECTING. This should happen when:

* `connect` is called ([RTN11c](https://docs.ably.io/client-lib-development-guide/features/#RTN11c)).
* `close` is called ([RTN12d](https://docs.ably.io/client-lib-development-guide/features/#RTN12d)).

This isn't captured by this change; `connect` causes a transition to CONNECTING, and `close` causes a transition to CLOSING. Maybe those were the states 

But there's a deeper issue here: the context is meant to be "live" for as along as we're _waiting_ for the timer, but it actually it can live longer than that. [After the timer fires](https://github.com/ably/ably-go/blob/43053b50d6f8e504c289df4ace37f46d195d2395/ably/realtime_conn.go#L344), this context should be dismissed, but it isn't; it will still be used on the next iteration of the loop if reconnection fails, and it will be _cancelled_ due to the transition to CONNECTING that happened while reconnecting, as you pointed out in the chat.

So my suggestion would be to move the body of the `for` loop in `connectWithRetryLoop` and  `connectAfterSuspension` to separate functions, and call `c.ctxCancelOnStateTransition()` there, so that the context is created afresh on each attempt to connect. (It would also make things simpler if `next` and `nextCtx` are moved inside this new function.)

Reply:
Okay, maybe there is confusion here. the purpose of the function is to stop reconnection when the state goes into the desired state. The primary state would be `CONNECTED` and not `CONNECTING`. The same goes for `CLOSED` and not `CLOSING`. There's no point in closing the timer when reconnection doesn't fulfill its purpose, that is to connect to the server

Reply:
@tcard, I have updated the reconnection logic, I think the only `connectionStateConnecting` wasn't getting updated at the right place. Everything else looks fine to me (including the creation of parent context)